[データサイエンスのためのPython入門14〜DataFrameの欠損値NaNに対応する〜](https://datawokagaku.com/dataframe_nan/)

こんにちは！あんちぽちゃんだよ〜〜〜！

今日もまた「データサイエンスのためのPython入門」をやっていきたいと思います！14回めの今回は「DataFrameの欠損値NaNに対応する」編！

「NaNって何だ？（笑）」（このネタ2回め）

ということで、やってみよ〜〜〜！

その前に、「NaNって何だ？」について真面目にみていこう。NaNってのは、Not a Numberの略ってのは第9回でやりましたね。

これからいろんなデータを扱っていく上で、NaNをどのように処理するかってのは適切な分析をする上で重要なことなのです！実際に扱うデータは、必ずしも全部のレコードに値がぎっしりつまっているとは限りません。実際、これまで使ってきたタイタニックのデータにもNaNがいっぱいあるんですね〜。

ではいつものやつ〜〜〜。

In [3]:
import numpy as np
import pandas as pd

In [4]:
df = pd.read_csv('data/titanic/train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

`Cabin`カラムにたくさんNaNがありますね〜。なんでなんだろ？客室がなくてそのへんに雑魚寝でもしてたのかな……。

というわけで、このNaNに対してどう扱っていくかってのを見ていきましょう！

まずは、NaNのあるレコードをdropしちゃう、`df.dropna()`！

In [5]:
df.dropna()

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

もともと891行あったデータが183行にまで減っちゃいましたね！ちなみに、第13回でやったフィルタリングのときと同じように、dropしたあともindexは維持されてますね。

データがたくさんある場合は`df.dropna()`でバーンと落としちゃってもいいかもですが、183行しか残らないようだと不安ですね……。

さて、`df.dropna()`に引数`axis=1`をわたすとNaNを含むカラムをdropできます！（他のメソッドと同じく、`axis=1`とするとカラムを表すんですね）。

In [6]:
df.dropna(axis = 1)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex  SibSp  Parch  \
0                              Braund, Mr. Owen Harris    male      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female      1      0   
2                               Heikkinen, Miss. Laina  female      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female      1      0   
4                             Allen, Mr. William Henry    male      0      0   
..                                                 ...     ...    ...    ...   
886                              Montvila, Rev. Juozas    male      0      0   
887                       Graham, Miss. Margaret Edith  female      0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female      1      2   
889                              Behr, Mr. Karl Howell    male      0      0   
890                                Dooley, Mr. Patrick    male      0      0   

               Ticket     Fare  
0           A/5 21171   7.2500  
1            PC 17599  71.2833  
2    STON/O2. 3101282   7.9250  
3              113803  53.1000  
4              373450   8.0500  
..                ...      ...  
886            211536  13.0000  
887            112053  30.0000  
888        W./C. 6607  23.4500  
889            111369  30.0000  
890            370376   7.7500  

[891 rows x 9 columns]

でもこれだと一個でもNaNがあると説明変数を削っちゃうことになるので、慎重に考えないとならないみたいです！（理由は、かめさんがのちに機械学習入門で教えてくれるそうです。楽しみですね！）

また、指定したカラムでNaNがあるレコードだけを落とす方法もあります。

In [7]:
df.dropna(subset=['Age'])

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

ちょっとわかりにくいですが、indexが888のレコードが、`Age`カラムの値がNaNだったのでdropされてます！

さて、これまではNaNなデータをdropしちゃうって方法を見てきましたが、なにか別の値を入れる方法もあります。

たとえば`df.fillna()`を使うと、一律に値をいれられます。

In [8]:
df.fillna('THIS IS IT!!1')

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex            Age  \
0                              Braund, Mr. Owen Harris    male             22   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female             38   
2                               Heikkinen, Miss. Laina  female             26   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female             35   
4                             Allen, Mr. William Henry    male             35   
..                                                 ...     ...            ...   
886                              Montvila, Rev. Juozas    male             27   
887                       Graham, Miss. Margaret Edith  female             19   
888           Johnston, Miss. Catherine Helen "Carrie"  female  THIS IS IT!!1   
889                              Behr, Mr. Karl Howell    male             26   
890                                Dooley, Mr. Patrick    male             32   

     SibSp  Parch            Ticket     Fare          Cabin Embarked  
0        1      0         A/5 21171   7.2500  THIS IS IT!!1        S  
1        1      0          PC 17599  71.2833            C85        C  
2        0      0  STON/O2. 3101282   7.9250  THIS IS IT!!1        S  
3        1      0            113803  53.1000           C123        S  
4        0      0            373450   8.0500  THIS IS IT!!1        S  
..     ...    ...               ...      ...            ...      ...  
886      0      0            211536  13.0000  THIS IS IT!!1        S  
887      0      0            112053  30.0000            B42        S  
888      1      2        W./C. 6607  23.4500  THIS IS IT!!1        S  
889      0      0            111369  30.0000           C148        C  
890      0      0            370376   7.7500  THIS IS IT!!1        Q  

[891 rows x 12 columns]

でも、実際の現場ではNaNを一律に何かにするみたいなことはないそうですね……。その代わり、何か代表値を入れることが多いんですね〜〜…たとえば、`Age`なら平均年齢とか。

まず、あるカラムの平均値はこうやって取ります。

In [9]:
df['Age'].mean()

29.69911764705882

ほんでもってこの平均値を`Age`カラムのNaNなとこに入れたいんだけど、どうするか？

まずはSeriesを作りましょう〜。

In [10]:
df['Age'].fillna(df['Age'].mean())

0      22.000000
1      38.000000
2      26.000000
3      35.000000
4      35.000000
         ...    
886    27.000000
887    19.000000
888    29.699118
889    26.000000
890    32.000000
Name: Age, Length: 891, dtype: float64

888の行に、さっき計算した平均年齢が入ってますね。ここは、元々NaNだったところです。そして、このSeriesを元のDataFrameに突っ込んでやると〜〜〜？

In [13]:
df['Age'] = df['Age'].fillna(df['Age'].mean())
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
0        1      0         A/5 21171   7.2500   NaN        S  
1        1      0          PC 17599  71.2833   C85        C  
2        0      0  STON/O2. 3101282   7.9250   NaN        S  
3        1      0            113803  53.1000  C123        S  
4        0      0            373450   8.0500   NaN        S  
..     ...    ...               ...      ...   ...      ...  
886      0      0            211536  13.0000   NaN        S  
887      0      0            112053  30.0000   B42        S  
888      1      2        W./C. 6607  23.4500   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

ジャジャーン！元のDataFrameの`Age`カラムが上書きされて、NaNだった888のレコードに平均年齢が入力されました〜〜〜。パチパチパチ。

さて、DataFrameの中の値がNaNかどうかを調べるにはどうしたらよいでしょうか？

NumPyには`np.isnan()`ってのがあるんですが、Pandasだと`pd.isna()`を使います（最後の`n`がないのがポイント）。

In [14]:
df = pd.read_csv('data/titanic/train.csv')
pd.isna(df)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

NaN判定の結果を別カラムで持ちたいときはこう！

In [15]:
df['Cabin_nam'] = pd.isna(df['Cabin'])
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Cabin_nam  
0        0         A/5 21171   7.2500   NaN        S       True  
1        0          PC 17599  71.2833   C85        C      False  
2        0  STON/O2. 3101282   7.9250   NaN        S       True  
3        0            113803  53.1000  C123        S      False  
4        0            373450   8.0500   NaN        S       True  
..     ...               ...      ...   ...      ...        ...  
886      0            211536  13.0000   NaN        S       True  
887      0            112053  30.0000   B42        S      False  
888      2        W./C. 6607  23.4500   NaN        S       True  
889      0            111369  30.0000  C148        C      False  
890      0            370376   7.7500   NaN        Q       True  

[891 rows x 13 columns]

一番右に、`Cabin_nan`というカラムができて、`True`/`False`が入りましたね〜。このように、新たにカラムを作りたいときは

`df['新しいカラム名'] = Series`とすると、もしそのカラムがあれば代入するし、なければ新規に作られるんですね。便利！

そんなわけで、今回は欠損値（NaN）をどう扱うかというテーマについてみてきました〜。ではまた！